File: heap.c
#include <stdio.h>
#include <stdlib.h>

typedef struct{
    int size;
    int* key;
    int* data; //on peut changer le type en fonction des données (exemple noeud pour Dijkstra)
} BinaryHeap;

void init_heap(BinaryHeap* heap, int heap_size){
    heap->size = 0;
    heap->key = malloc(heap_size * sizeof(int));
    heap->data = malloc(heap_size * sizeof(int));
}

void add(BinaryHeap* heap, int key, int data){
    //ajoute au tas un element
    heap->size++;
    heap->key[heap->size - 1] = key;
    heap->data[heap->size - 1] = data;

    if(heap->size != 1){
        int i = heap->size - 1;
        int changed = 1;
        while(changed){
            changed = 0;
            //si on est à la racine
            if(i == 0) break;
            else if (heap->key[(i-1) >> 1] > heap->key[i]){ 
                //swap
                int a = heap->key[(i-1) >> 1];
                int b = heap->key[i];
                heap->key[(i-1) >> 1] = b;
                heap->key[i] = a;
                int c = heap->data[(i-1) >> 1];
                int d = heap->data[i];
                heap->data[(i-1) >> 1] = d;
                heap->data[i] = c;
                //on passe au parent
                i = ((i-1) >> 1);
                changed = 1;
            }
        }
    }
}

void delete(BinaryHeap* heap){
    //enlève le premier element du tas
    if(heap->size > 0){
        heap->key[0] = heap->key[heap->size -1];
        heap->data[0] = heap->data[heap->size -1];
        heap->size--;
        int i = 0;
        int changed = 1;
        while(changed){
            changed = 0;

            int left = 2 * i + 1;
            int right = 2 * i + 2;
            int smallest_child = left;
            if(left > heap->size - 1) break;
            
            if(heap->size > right){ 
                if(heap->key[left] > heap->key[right]){ 
                    smallest_child = right;
                }
            }
            if(heap->key[smallest_child] < heap->key[i]){ 
                //swap
                int a = heap->key[smallest_child];
                int b = heap->key[i];
                heap->key[smallest_child] = b;
                heap->key[i] = a;
                int c = heap->data[smallest_child];
                int d = heap->data[i];
                heap->data[smallest_child] = d;
                heap->data[i] = c;
                //on passe à l'enfant
                i = smallest_child;
                changed = 1;
            }
        }
    }
}

void print_heap(BinaryHeap* heap){

    for(int i = 0; i < heap->size; i++){
        printf("%d ", heap->key[i]);
    }
    printf("\n");
}

int main(){

    BinaryHeap heap;
    init_heap(&heap, 1000);
    add(&heap, 1, 2);
    add(&heap, 4, 1);
    add(&heap, 2, 2);
    add(&heap, 3, 4);
    print_heap(&heap);
    delete(&heap);
    print_heap(&heap);

    return 0;
}

File: Heap.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Feb 23 17:43:24 2026

@author: victordebanne
"""


class BinaryHeap():
    def __init__(self):
        self.heap = []
        self.size = 0
        #parent de i est (i-1) >> 1
        #enfant de i est (i-1) << 1
        
    def add(self, key, data):
        self.heap.append((key, data))
        self.size += 1
        if self.size != 1 : 
            i = self.size - 1
            changed = True 
            while changed : 
                changed = False
                #si on est à la racine
                if i == 0 : 
                    break
                if self.heap[(i-1) >> 1][0] > self.heap[i][0] : 
                    #swap
                    self.heap[(i-1) >> 1], self.heap[i] = self.heap[i], self.heap[(i-1) >> 1]
                    #on passe au parent
                    i = ((i-1) >> 1)
                    changed = True
                               
    def remove(self):
        if self.size > 0 : 
            self.heap[0] = self.heap[-1]
            self.heap.pop(-1)
            self.size -= 1
            i = 0
            changed = True 
            while changed : 
                changed = False
                left = 2 * i + 1
                right = 2 * i + 2
                smallest_child = left
                if right > self.heapsize - 1 : 
                    break
                if self.size > right : 
                    if self.heap[left][0] > self.heap[right][0] : 
                        smallest_child = right
    
                if self.heap[smallest_child][0] < self.heap[i][0] : 
                    #swap
                    self.heap[smallest_child], self.heap[i] = self.heap[i], self.heap[smallest_child]
                    #on passe à l'enfant
                    i = smallest_child
                    changed = True
            
        
    
if __name__ == "__main__":
    heap = BinaryHeap()
    heap.add(1, "bonjour")
    heap.add(4, "au revoir")
    heap.add(2, "ça va")
    heap.add(3, "et toi")
    
    print(heap.heap)
    
    


        
        
        